automation/emit_coverage_info.sh

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env bash
+
+# Run tests in order to produce coverage data. Does not collect coverage data,
+# just produces it. Note that we ignore all test failures -- some of the
+# experimental flags we use cause test failures, we don't really care, other CI
+# tasks check this anyway.
+#
+# Note that running this outside of CI doesn't work on all platforms. In
+# particular, macos needs further patches in order to work, probably windows
+# too. It's really intended for use inside CI only.
+#
+# Finally, this must be done on nightly -- `cargo-tarpaulin`, which works on
+# stable, also has issues with NSS, but not ones that seem easy to work around.
+# As a result, we use `grcov`. Ultimately `grcov` should long-term allow us to
+# integrate coverage data from over the FFI, which is a huge plus to this
+# approach, so it's probably for the best, even if `grcov`'s actual coverage
+# information seems to be of lower quality than `tarpaulin`.
+
+if [[ ! -f "$PWD/automation/emit_coverage_info.sh" ]]
+then
+    echo "emit_coverage_info.sh must be executed from the root directory."
+    exit 1
+fi
+
+cargo +nightly clean
+# Note that incremental builds, as well as sccache seems to break coverage :(.
+# So, we disable those.
+export CARGO_INCREMENTAL=0
+unset RUSTC_WRAPPER
+# TODO: re-add `-Clink-dead-code` to RUSTFLAGS -- right now it causes a build
+# linker error in NSS, as apparently some part of NSS we don't use doesn't have
+# the build flags set up correctly. This makes our code coverage accuracy worse,
```

Comment:
> as apparently some part of NSS we don't use doesn't have the build flags set up correctly

Do we know which parts? I wouldn't mind opening an issue for that.

Reply:
This is the linker error:
```
  = note: /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_aes_gcmInitCounter':
          intel-gcm-wrap.c:(.text.intel_aes_gcmInitCounter+0x3af): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_aes_gcmInitCounter+0x3d5): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_CreateContext':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_CreateContext+0x19): undefined reference to `PORT_ZAlloc_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_CreateContext+0xb6): undefined reference to `PORT_Free_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DestroyContext':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DestroyContext+0x44): undefined reference to `PORT_Free_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_EncryptUpdate':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0xed): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0xfe): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptUpdate+0x10f): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DecryptUpdate':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xa8): undefined reference to `NSS_SecureMemcmp_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xd1): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xe2): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0xf3): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptUpdate+0x119): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_EncryptAEAD':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptAEAD+0x144): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o):intel-gcm-wrap.c:(.text.intel_AES_GCM_EncryptAEAD+0x156): more undefined references to `PORT_SetError_stub' follow
          /usr/bin/ld: /home/circleci/project/libs/desktop/linux-x86-64/nss/lib/libintel-gcm-wrap_c_lib.a(intel-gcm-wrap_c_lib.intel-gcm-wrap.o): in function `intel_AES_GCM_DecryptAEAD':
          intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0xdb): undefined reference to `NSS_SecureMemcmp_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x11a): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x12b): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x13c): undefined reference to `PORT_SetError_stub'
          /usr/bin/ld: intel-gcm-wrap.c:(.text.intel_AES_GCM_DecryptAEAD+0x162): undefined reference to `PORT_SetError_stub'
          collect2: error: ld returned 1 exit status
```

Never managed to trigger it on my linux laptop -- only in CI.

Reply:
Filed https://github.com/mozilla/application-services/issues/3390